File: python/algos/minimum_depth_of_bst.py
from __future__ import annotations


class TreeNode:
    def __init__(self, value: int = 0, left: TreeNode | None = None, right: TreeNode | None = None) -> None:
        self.val = value
        self.left = left
        self.right = right


def minimum_depth_of_bst(node: TreeNode | None) -> int:
    if node is None:
        return 0
    elif node.left is None and node.right is None:
        return 1
    elif node.left is None:
        return 1 + minimum_depth_of_bst(node.right)
    elif node.right is None:
        return 1 + minimum_depth_of_bst(node.left)
    else:
        return 1 + min(minimum_depth_of_bst(node.left), minimum_depth_of_bst(node.right))


if __name__ == '__main__':
    assert minimum_depth_of_bst(TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15, TreeNode(8)), TreeNode(7)))) == 2

    assert minimum_depth_of_bst(TreeNode(1, None, TreeNode(2, None, TreeNode(3, None, TreeNode(4, None, TreeNode(5)))))) == 5
    assert minimum_depth_of_bst(TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3))) == 2

File: rust/src/algos/minimum_depth_of_bst.rs
use std::cell::RefCell;
use std::cmp::min;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub fn minimum_depth_of_bst(node: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    match node {
        Some(node) => match (node.borrow().left.clone(), node.borrow().right.clone()) {
            (Some(left), Some(right)) => {
                1 + min(
                    minimum_depth_of_bst(node.borrow().left.clone()),
                    minimum_depth_of_bst(node.borrow().right.clone()),
                )
            }
            (Some(left), None) => 1 + minimum_depth_of_bst(node.borrow().left.clone()),
            (None, Some(right)) => 1 + minimum_depth_of_bst(node.borrow().right.clone()),
            (None, None) => 1,
        },
        None => 0,
    }
}
